### PR TITLE
MADlib image loader:  load dataset from disk

### PR DESCRIPTION
**Pull request by reductionista**
Adds to the madlib_image_loader module ability to load images
directly from disk (instead of having to be in numpy arrays).

New functionality can either be called as a module or run from
the command line.  (Loading from a numpy array still only
supported when called as a module.)

Dataset to be loaded is expected to be organized by image
category/class.  Each directory name labels a specific class,
and only images from that class should be in that directory.

An additional img_name column is added to the table when used in
this way, to help with tracking which image is which later.

Madlib Image Loader Demo notebook has been updated to include a
minor interface change (original API endpoint renamed to
load_dataset_from_np) and an example of loading from disk with
newly added endpoint load_dataset_from_disk.

### CODE REVIEW

File: community-artifacts/Deep-learning/madlib_image_loader.py [hunk 1]
```diff
@@ -383,36 +495,180 @@ def load_np_array_to_table(self, data_x, data_y, table_name=None,
...
         #
-        #  multiprocessing library will call _call_worker() in some worker for
+        #  multiprocessing library will call _call_np_worker() in some worker for
         #   each file, splitting the list of files up into roughly equal chunks
         #   for each worker to handle.  For example, if there are 500 files and
-        #   5 workers, each will handle about 100 files, and _call_worker() will
-        #   be called 100 times, each time with a different file full of images.
-        #
+        #   5 workers, each will handle about 100 files, and _call_np_worker()
+        #   will #   be called 100 times, each time with a different file full
```
Comment (orhankislal) [inline community-artifacts/Deep-learning/madlib_image_loader.py]: will #   be -> will be

**Review by orhankislal — APPROVED** [review]
LGTM

**Comment by fmcquillan99** [comment]
which formats of image files do we support?  @reductionista

**Comment by fmcquillan99** [comment]
If number of examples is less than `ROWS_PER_FILE` then no rows get written to the target table, it seems.

```
%sql drop table if exists load_train_1;
# Load images from file system
iloader.load_dataset_from_disk('/Users/fmcquillan/tmp/cifar-10-png-sample1', 'load_train_1', num_labels='all', append=False, no_temp_files=False)
```
produces
```
Done.
MainProcess: Connected to madlib db.
Executing: CREATE TABLE load_train_1 (id SERIAL, x REAL[], y TEXT,                        img_name TEXT)
CREATE TABLE
Created table load_train_1 in madlib db
.DS_Store is not a directory, skipping
number of labels = 3
Found 3 image labels in /Users/fmcquillan/tmp/cifar-10-png-sample1
Spawning 5 workers...
Initializing PoolWorker-1 [pid 48064]
Initializing PoolWorker-2 [pid 48065]
PoolWorker-1: Created temporary directory /tmp/madlib_lOZkc8k825
Initializing PoolWorker-3 [pid 48066]
PoolWorker-2: Created temporary directory /tmp/madlib_B8aEPagCVQ
PoolWorker-3: Created temporary directory /tmp/madlib_dbgZdbFCkG
Initializing PoolWorker-4 [pid 48067]
Initializing PoolWorker-5 [pid 48068]
PoolWorker-4: Created temporary directory /tmp/madlib_opXnfyKRjF
PoolWorker-5: Created temporary directory /tmp/madlib_YLxwFljDy1
PoolWorker-4: Connected to madlib db.
PoolWorker-2: Connected to madlib db.
PoolWorker-1: Connected to madlib db.
PoolWorker-5: Connected to madlib db.
PoolWorker-3: Connected to madlib db.
filenames ['35097_truck.png', '35116_truck.png', '35127_truck.png']
filenames ['35096_bird.png', '35104_bird.png', '35107_bird.png']
index 0 filename 35096_bird.png
index 1 filename 35104_bird.png
index 2 filename 35107_bird.png
filenames ['35100_cat.png', '35101_cat.png', '35102_cat.png']
index 0 filename 35097_truck.png
index 0 filename 35100_cat.png
index 1 filename 35116_truck.png
index 1 filename 35101_cat.png
index 2 filename 35127_truck.png
index 2 filename 35102_cat.png
PoolWorker-5: Removed temporary directory /tmp/madlib_YLxwFljDy1
PoolWorker-3: Removed temporary directory /tmp/madlib_dbgZdbFCkG
PoolWorker-4: Removed temporary directory /tmp/madlib_opXnfyKRjF
PoolWorker-2: Removed temporary directory /tmp/madlib_B8aEPagCVQ
PoolWorker-1: Removed temporary directory /tmp/madlib_lOZkc8k825
Done!  Loaded 3 image categories in 0.356377124786s
5 workers terminated.
```
and get empty table
```
madlib=# select count(*) from load_train_1;
 count 
-------
     0
(1 row)
```

**Comment by fmcquillan99** [comment]
This might be OK, but wondering why 

(1)
if I load from filesystem:
```
%sql drop table if exists cifar_10_train_data_filesystem:
# Load images from file system
iloader.load_dataset_from_disk('/Users/fmcquillan/tmp/cifar10', 'cifar_10_train_datafile_system', num_labels='all', append=False, no_temp_files=False)
```
I get this format:
```
cifar_10_train_data filesystem
{{{{218,239,244},{219,240,245},{220,241,246}, ...
```

(2)
and if I load from numpy: 
```
# Load dataset into np array
(x_train, y_train), (x_test, y_test) = cifar10.load_data()

%sql DROP TABLE IF EXISTS cifar_10_train_data, cifar_10_test_data;

# Save images to temporary directories and load into database
iloader.load_dataset_from_np(x_train, y_train, 'cifar_10_train_data', append=False, no_temp_files=False)
iloader.load_dataset_from_np(x_test, y_test, 'cifar_10_test_data', append=False, no_temp_files=False)
```
I get this format:
```
cifar_10_train_data
{{{59,62,63},{43,46,45},{50,48,43}, ...
```

It means if I run the model architecture from https://keras.io/examples/cifar10_cnn/ it works when loading from numpy (2) but when I run from on filesystem loaded data (1) I get this message:
```
ERROR:  plpy.Error: model_keras error: Input shape [32, 32, 3] in the model architecture does not match the input shape [1, 32, 32] of column independent_var in table cifar_10_train_data_packed.
CONTEXT:  Traceback (most recent call last):
  PL/Python function "madlib_keras_fit", line 21, in <module>
    madlib_keras.fit(**globals())
  PL/Python function "madlib_keras_fit", line 42, in wrapper
  PL/Python function "madlib_keras_fit", line 102, in fit
  PL/Python function "madlib_keras_fit", line 308, in validate_input_shapes
  PL/Python function "madlib_keras_fit", line 86, in _validate_input_shapes
PL/Python function "madlib_keras_fit"
```

**Comment by fmcquillan99** [comment]
Now loading from file system it matches numpy
```
{{{26,30,33},{20,24,27},{16,20,23}...
```
Then I mini-batched it and ran keras fit() and it went though OK.

So thanks for making the changes.

LGTM

**Pull request closed without merging**